File: 368.largest-divisible-subset.py
#
# @lc app=leetcode id=368 lang=python3
#
# [368] Largest Divisible Subset
#

# Time:  O(n^2)
# Space: O(n)

from typing import List


# @lc code=start
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        nums.sort()
        n, max_idx = len(nums), 0
        dp, prev = [1] * n, [-1] * n
        for i in range(n):
            for j in range(i):
                if nums[i] % nums[j] == 0:
                    if dp[i] < dp[j] + 1:
                        dp[i] = dp[j] + 1
                        prev[i] = j
            if dp[max_idx] < dp[i]:
                max_idx = i

        res, i = [], max_idx
        while i != -1:
            res.append(nums[i])
            i = prev[i]
        return res[::-1]


# @lc code=end

File: 368.largest-divisible-subset.js
/*
 * @lc app=leetcode id=368 lang=javascript
 *
 * [368] Largest Divisible Subset
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var largestDivisibleSubset = function (nums) {
  nums.sort((a, b) => a - b);
  let maxIndex = 0;
  const dp = Array(nums.length).fill(1);
  const prev = Array(nums.length).fill(-1);
  for (let i = 0; i < nums.length; ++i) {
    for (let j = 0; j < i; ++j) {
      if (nums[i] % nums[j] === 0) {
        if (dp[i] < dp[j] + 1) {
          dp[i] = dp[j] + 1;
          prev[i] = j;
        }
      }
    }
    if (dp[maxIndex] < dp[i]) {
      maxIndex = i;
    }
  }
  const res = [];
  for (let i = maxIndex; i !== -1; i = prev[i]) {
    res.push(nums[i]);
  }
  return res.reverse();
};
// @lc code=end
